### PR TITLE
build: add a CMake based build for Atomics

### PR DESCRIPTION
**Pull request by compnerd**
This adds an initial implementation of a CMake based build system.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @tomerd

**Comment by compnerd** [comment]
This should allow building with CMake, but this is insufficient for building on Windows.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#[[
+This source file is part of the Swift Atomics Open Source Project
+
+Copyright (c) 2020 Apple Inc. and the Swift Numerics project authors
```
Comment (lorentey) [inline CMakeLists.txt:4]: For consistency, I believe this should read:

```suggestion
Copyright (c) 2021 Apple Inc. and the Swift project authors
```

File: Sources/Atomics/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#[[
+This source file is part of the Swift Atomics Open Source Project
+
+Copyright (c) 2020 Apple Inc. and the Swift Numerics project authors
```
Comment (lorentey) [inline Sources/Atomics/CMakeLists.txt:4]:
```suggestion
Copyright (c) 2021 Apple Inc. and the Swift project authors
```

File: Sources/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#[[
+This source file is part of the Swift Atomics Open Source Project
+
+Copyright (c) 2020 Apple Inc. and the Swift Numerics project authors
```
Comment (lorentey) [inline Sources/CMakeLists.txt:4]:
```suggestion
Copyright (c) 2021 Apple Inc. and the Swift project authors
```

File: Sources/_AtomicsShims/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#[[
+This source file is part of the Swift Atomics Open Source Project
+
+Copyright (c) 2020 Apple Inc. and the Swift Numerics project authors
```
Comment (lorentey) [inline Sources/_AtomicsShims/CMakeLists.txt:4]:
```suggestion
Copyright (c) 2021 Apple Inc. and the Swift project authors
```

File: cmake/modules/SwiftSupport.cmake [hunk 5]
```diff
@@ -0,0 +1,94 @@
+#[[
+This source file is part of the Swift Atomics Open Source Project
+
+Copyright (c) 2019 Apple Inc. and the Swift Numerics project authors
```
Comment (lorentey) [inline cmake/modules/SwiftSupport.cmake:4]:
```suggestion
Copyright (c) 2021 Apple Inc. and the Swift project authors
```

**Review by lorentey — APPROVED** [review]
Thank you! 

We'll need to set up CI support for building the library for these. I don't suppose we can run the tests on the resulting binary, can we?

**Comment by lorentey** [comment]
@swift-ci test

**Comment by compnerd** [comment]
I can look at adding the necessary build infrastructure for tests to this.  That isn't too bad to do, but getting this package building on Windows is more challenging.  I ran into problems with the being unable to support the needed code generation.

**Comment by lorentey** [comment]
Ah -- tests would be nice to run, but making it build is definitely more important.

**Comment by compnerd** [comment]
@lorentey - okay, I have this building on Windows now.  I'm going to look into the testing situation with CMake as well.

**Comment by compnerd** [comment]
Okay, once #29 is merged, I'll upload a subsequent change to this to enable the tests.  With these two patches, I can build and pass the test suite on Windows with CMake.  The build with SPM works, but does not generate a shared library, and static linking on Windows is currently not supported.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
While I cannot help figure out how to get this integrated into the swift CI system (@tomerd and @shahmishal should be able to help with that), I can offer a daily run and occasional one-offs if needed via Azure: https://dev.azure.com/compnerd/swift-build/_build/results?buildId=50359&view=results

**Pull request merged**